lib/navigation.php

```diff
@@ -571,3 +571,186 @@ function gutenberg_get_navigation_areas_paths_to_preload() {
 	}
 	return $paths;
 }
+
+/**
+ * Migrates classic menus to block-based menus on theme switch.
+ *
+ * @param string   $new_name  Name of the new theme.
+ * @param WP_Theme $new_theme New theme.
+ * @param WP_Theme $old_theme Old theme.
+ * @see switch_theme WordPress action.
+ */
+function gutenberg_migrate_nav_on_theme_switch( $new_name, $new_theme, $old_theme ) {
+	// Do nothing when switching to a theme that does not support site editor.
+	if ( ! gutenberg_experimental_is_site_editor_available() ) {
+		return;
+	}
+	global $wpdb;
+
+	// get_nav_menu_locations() calls get_theme_mod() which depends on the stylesheet option.
+	// At the same time, switch_theme runs only after the stylesheet option was updated to $new_theme.
+	// To retrieve theme mods of the old theme, let's pretend the stylesheet is as it used to be.
+	$pretend_old_theme = function() use ( $old_theme ) {
```

Comment:
What does pretend mean? It obfuscates the simplicity here: we use the option_* filter to filter the output of the stylesheet option to get the value as if it were the old theme, because `switch_theme` gives old theme to this callback.